### ISSUE TITLE
Transformer Training Endpoints Fit Production Models on Random Synthetic Noise, Silently Destroying Forecast Quality

### ISSUE THREAD

**Issue by ionfwsrijan** [KanishJebaMathewM/Truxify#11239]
# Issue: Transformer Training Endpoints Fit Production Models on Random Synthetic Noise, Silently Destroying Forecast Quality

## Description

`backend/ml/routes/transformer_routes.py` exposes `/transformer/{demand,traffic,price}/train`. Each `TrainRequest` is served by a **single global trainer that owns the exact model instance used by the live forecast endpoints**:

```python
# module level
demand_model = DemandForecastTransformer()
traffic_model = TrafficForecastTransformer()
price_model = PriceForecastTransformer()
demand_trainer = TransformerTrainer(demand_model)
traffic_trainer = TransformerTrainer(traffic_model)
price_trainer = TransformerTrainer(price_model)
```

The training endpoints then fit those *same* shared model objects on **synthetic random noise** (`torch.randn`):

```python
@router.post(/demand/train)
async def train_demand(request: TrainRequest):
    train_data = torch.randn(1000, demand_model.transformer.seq_len, demand_model.input_dim)  # pure noise
    train_labels = torch.randn(1000, demand_model.transformer.pred_len)                        # pure noise
    ...
    results = demand_trainer.train(train_data, train_labels, epochs=request.epochs, ...)
    return {'success': True, 'data': results, ...}
```

Because `TransformerTrainer` holds a reference to the same `nn.Module` that `forecast_demand` runs inference on, each `/train` call **overwrites the production model weights** in place via `train_step` → `optimizer.step()`. `forecast_demand` runs `demand_trainer.predict` against the *post-train* (now noise-fitted) weights.

**Critical problems**:

1. **Weights replaced by a noise-fitted model.** Training against `y = randn()` maximizes (or randomizes) the loss surface with zero signal. Any earlier trained/validated weights are destroyed the moment someone calls `/demand/train`. There is no separation between the training artifact and the serving artifact — they are literally the same tensor buffers.

2. **`train()` has no `torch.no_grad()` isolation and returns `avg_loss` unconditionally.** `TransformerTrainer.train` (model.py:282-328) computes `num_batches = (train_data.size(0)+batch_size-1)//batch_size` and divides by it; with the empty-data branch (`train_data.size(0)==0 → batch loop never runs`) `len` divides by zero → `ZeroDivisionError`, and with noise it trains destructively.

3. **No locking / no versioning.** `train` and `forecast` hitting the same shared objects concurrently race: `train_step` calls `self.model.train()` (enables dropout) and mutates weights while `predict` is mid-forward. Forecasting under concurrent `/train` returns nondeterministic, corrupted outputs (see ISSUE-015's global-state class — but here the *trainer holds the serving model*, so even sequential calls irreversibly degrade serving).

4. **No auth / parameter bounds.** `epochs` and `batch_size` are taken straight from the request body. `epochs: 1e6`, `batch_size: 1`, or negative values are accepted; a huge `epochs` triggers a CPU/GPU-heavy burn, and `batch_size: 0` → division by zero in `num_batches`.

5. **Three endpoints, zero validation, zero tests.** All three (`demand`, `traffic`, `price`) share the identical destructively-trained pattern.

## Expected Behavior

- Training must write to a *separate* model artifact (new weights / new `TransformerTrainer`), never mutate the serving model. Serving stays frozen on the last validated checkpoint.
- `/train` must validate `0 < epochs` and `batch_size >= 1`, require auth, and be gated by an exclusive train lock (or run in a worker).
- After training, a new model version should be atomically swapped into serving; `/forecast` should be decoupled from training.

## Actual Behavior

- One `/demand/train` request permanently replaces production demand-model weights with a model fit on uniform random noise — forecasts for every subsequent request are garbage until someone manually re-deploys/re-trains on real data.
- Concurrent `/train` + `/forecast` produce racy nondeterministic predictions and can throw mid-forward errors (`self.model.train()` dropout / weight mutation during eval).
- `batch_size: 0` → `ZeroDivisionError`; huge `epochs` → resource exhaustion; all under no auth.

## Proposed Fix

1. Give the trainer its *own* model copy for training and a swap callback:
   ```python
   def train(self, ...):
       working = copy.deepcopy(self.model)
       ... train working ...
       self.model.load_state_dict(working.state_dict())  # atomic swap at end only
   ```
   or keep serving on `model_path`-loaded weights and only swap on successful validation (early-stop on `val_loss`).
2. Add Pydantic constraints on `TrainRequest` (`epochs: int = Field(..., ge=1, le=500)`, `batch_size: int = Field(..., ge=1, le=1024)`) and guard `num_batches` against zero.
3. Add an `asyncio.Lock` / worker queue around the `*_train` handlers so training and forecasting never interleave on the same tensors; enforce `torch.no_grad()` in serving `predict`.
4. Add auth + audit logging to the training endpoints, and a `/model/version` endpoint exposing the currently-serving version.
5. Tests: assert forecast is unchanged during training, then changes only after an explicit commit; assert `epochs=0`/`batch_size=0` rejected.

## Files to Modify

- `backend/ml/routes/transformer_routes.py` — bound training params, auth, locking, artifact swap.
- `backend/ml/transformers/model.py` — safe `num_batches`, immutable-train pattern, `no_grad` predict.
- `backend/ml/routes/mlRoutes.js` (`repo/backend/api/...` if it proxies `/transformer`) — ensure it forwards auth/session.

## Severity

**Critical** — a single unauthenticated request silently destroys the live forecasting models; all downstream ETA/pricing features consume garbage afterwards with no indication of corruption.

**Comment by kazi716** [comment]
### Issue Assignment Request (L3 - Core/Architecture/Performance & Security Hardening)

I would like to work on this issue. Based on my analysis, this task qualifies as an L3 task (Core/Architecture/Performance) because it addresses a critical architectural flaw and data corruption vulnerability in the ML serving pipeline where training mutates production weights in-place.

**Implementation Plan:**
1. **Modify \ackend/ml/transformers/model.py\**:
   - **Immutable Training & Atomic Swap**: Refactor \TransformerTrainer.train()\ to perform training on a deep copy of the model (\copy.deepcopy(self.model)\) and only swap the weights (\load_state_dict\) via an atomic update upon successful completion.
   - **Concurrency Safety**: Implement an \syncio.Lock\ around the training routine to ensure thread-safe separation between training and concurrent forecasting.
   - **Inference Optimization**: Enforce \	orch.no_grad()\ and \.eval()\ within the prediction path to prevent accidental graph accumulation and dropout execution.
   - **Math Safety**: Guard \atch_size\ and \
um_batches\ calculations against \ZeroDivisionError\.
2. **Modify \ackend/ml/routes/transformer_routes.py\**:
   - **Input Validation Boundaries**: Introduce strict Pydantic constraints (\Field(ge=1, le=500)\ for \pochs\, \atch_size\, etc.) in \TrainRequest\.
   - **Data Pipeline Integrity**: Eliminate the use of \	orch.randn\ synthetic noise, forcing the endpoints to accept real data configurations.
3. **Testing & Validation**:
   - Write regression tests to verify that \predict()\ outputs remain deterministic during a concurrent training run and that zero/negative parameters are properly rejected by Pydantic.

Please assign this issue to me, and I will prepare a PR with these architectural and security fixes.
